slang-gfx.h

```diff
@@ -1183,6 +1178,266 @@ struct WindowHandle
+    virtual SLANG_NO_THROW void SLANG_MCALL
+        drawIndexed(UInt indexCount, UInt startIndex = 0, UInt baseVertex = 0) = 0;
+    virtual SLANG_NO_THROW void SLANG_MCALL setStencilReference(uint32_t referenceValue) = 0;
+};
+#define SLANG_UUID_IRenderCommandEncoder                                               \
+    {                                                                                  \
+        0x39417cf7, 0x8d97, 0x43a9, { 0xbb, 0x9f, 0x2f, 0x35, 0xe9, 0x11, 0xd0, 0x42 } \
+    }
+
+class IComputeCommandEncoder : public ICommandEncoder
+{
+public:
+    virtual SLANG_NO_THROW void SLANG_MCALL
+        bindRootShaderObject(PipelineType pipelineType, IShaderObject* object) = 0;
+
+    virtual SLANG_NO_THROW void SLANG_MCALL setDescriptorSet(
+        PipelineType pipelineType,
+        IPipelineLayout* layout,
+        UInt index,
+        IDescriptorSet* descriptorSet) = 0;
+
+    virtual SLANG_NO_THROW void SLANG_MCALL setPipelineState(IPipelineState* state) = 0;
+    virtual SLANG_NO_THROW void SLANG_MCALL dispatchCompute(int x, int y, int z) = 0;
+};
+#define SLANG_UUID_IComputeCommandEncoder                                              \
+    {                                                                                  \
+        0x65400452, 0xc877, 0x478f, { 0x91, 0x7d, 0x48, 0xd5, 0x41, 0x6f, 0x39, 0xab } \
+    }
+
+class IResourceCommandEncoder : public ICommandEncoder
+{
+public:
+    virtual SLANG_NO_THROW void SLANG_MCALL copyBuffer(
+        IBufferResource* dst,
+        size_t dstOffset,
+        IBufferResource* src,
+        size_t srcOffset,
+        size_t size) = 0;
+    virtual SLANG_NO_THROW void SLANG_MCALL
+        uploadBufferData(IBufferResource* dst, size_t offset, size_t size, void* data) = 0;
+};
+#define SLANG_UUID_IResourceCommandEncoder                                             \
+    {                                                                                  \
+        0x5fe87643, 0x7ad7, 0x4177, { 0x8b, 0xd1, 0xd7, 0x84, 0xad, 0xcf, 0x3d, 0xce } \
+    }
+
+class ICommandBuffer : public ISlangUnknown
+{
+public:
+    // Only one encoder may be open at a time. User must call `ICommandEncoder::endEncoding`
+    // before calling other `encode*Commands` methods.
+    // Once `endEncoding` is called, the `ICommandEncoder` object becomes obsolete and is
+    // invalid for further use. To continue recording, the user must request a new encoder
+    // object by calling one of the `encode*Commands` methods again.
+    virtual SLANG_NO_THROW void SLANG_MCALL encodeRenderCommands(
+        IRenderPassLayout* renderPass,
+        IFramebuffer* framebuffer,
+        IRenderCommandEncoder** outEncoder) = 0;
+    ComPtr<IRenderCommandEncoder>
+        encodeRenderCommands(IRenderPassLayout* renderPass, IFramebuffer* framebuffer)
+    {
+        ComPtr<IRenderCommandEncoder> result;
+        encodeRenderCommands(renderPass, framebuffer, result.writeRef());
```

Comment:
This is in no way a criticism of this PR, but we should eventually decide whether the interfaces in `gfx` should contain these convenience operations or not, and if we *do* keep them, then we probably need to change them so that they do something with the `Result`s of the operations they call (something like wrap the call in `GFX_CHECK_RESULT` or similar, so that the user can define that macro to be something that `throw`s or `assert`s).

I notice in this case the operations return `void`, and I don't know if that is intentional, because it seems like a command buffer created for a compute-only renderer/device/queue should be able to fail on `encodeRenderCommands()`.